core/server/master/src/main/java/alluxio/master/file/contexts/OperationContext.java

```diff
@@ -11,54 +11,89 @@
+import java.util.LinkedList;
+import java.util.List;
 
 /**
  * Used as a base class for wrapping context around proto messages.
  *
  * @param <T> Proto message type
+ * @param <C> extended type
  */
 @NotThreadSafe
-public class OperationContext<T extends com.google.protobuf.GeneratedMessageV3.Builder<?>> {
+public class OperationContext<T extends GeneratedMessageV3.Builder, C extends OperationContext> {
   // Proto message that is being wrapped
   private T mOptionsBuilder;
   // Used to track client call status.
-  private CallTracker mCallTracker;
+  private List<CallTracker> mCallTrackers;
 
   /**
    * Creates an instance with given proto message.
    *
    * @param optionsBuilder Internal proto message builder instance
    */
   public OperationContext(T optionsBuilder) {
-    this(optionsBuilder, null);
     mOptionsBuilder = optionsBuilder;
-    mCallTracker = CallTracker.NOOP_TRACKER;
+    mCallTrackers = new LinkedList<>();
   }
 
   /**
-   * Creates an instance with given proto message.
+   * @return underlying proto message instance
+   */
+  public T getOptions() {
+    return mOptionsBuilder;
+  }
+
+  /**
+   * Used to transfer trackers of a context to this instance.
+   * This is required when creating internal contexts.
    *
-   * @param optionsBuilder Internal proto message builder instance
-   * @param callTracker client call tracker, or {@code null} if no tracking is desired
+   * @param context the source context
+   * @return the updated instance
    */
-  public OperationContext(T optionsBuilder, CallTracker callTracker) {
-    mOptionsBuilder = optionsBuilder;
-    mCallTracker = callTracker;
+  public C withTracker(OperationContext context) {
+    mCallTrackers = context.mCallTrackers;
+    return (C) this;
   }
 
   /**
-   * @return underlying proto message instance
+   * Updates this context with a new tracker.
+   *
+   * @param tracker the new call tracker
+   * @return the updated instance
    */
-  public T getOptions() {
-    return mOptionsBuilder;
+  public C withTracker(CallTracker tracker) {
+    mCallTrackers.add(tracker);
+    return (C) this;
   }
 
   /**
-   * TODO(ggezer): Make the call-tracker infra note the source of cancellation.
-   * @return {@code true} if the call is cancelled by the client
+   * @return the list of trackers that have cancelled this operation
    */
-  public boolean isCancelled() {
-    return mCallTracker.isCancelled();
+  public List<CallTracker> getCancelledTrackers() {
+    boolean trackerCancelled = false;
+    for (CallTracker tracker : mCallTrackers) {
+      if (tracker.isCancelled()) {
+        trackerCancelled = true;
+        break;
+      }
+    }
+
+    if (!trackerCancelled) {
+      return Collections.emptyList();
+    }
+
+    List<CallTracker> cancelledTrackers = new LinkedList<>();
+    for (CallTracker tracker : mCallTrackers) {
+      if (tracker.isCancelled()) {
+        cancelledTrackers.add(tracker);
+      }
+    }
+
+    return cancelledTrackers;
```

Comment:
The list of call trackers should always be small, correct? 1 most of the time?

Are there any cases when we could have more than that?

Reply:
Every context has a state-lock tracker. Some has an additional transport level tracker (GrpcCallTracker).
Even for those, it's very unlikely to have more than 1 canceller. However, it's possible so I give out the list here. (We might also wanna add more trackers in the future...)